File: 4KYU/Next bigger number with the same digits/Solution.cs
using System.Collections.Generic;
using System.Linq;
using System;

public class Kata
{
    public static long NextBiggerNumber(long n)
    {
        var digits = n.ToDigits().ToArray();
        var d = -1;

        for (var i = digits.Length-2; i >= 0; i--)
        {
            if (digits[i] >= digits[i + 1])
                continue;
            d = i;
            break;
        }

        if (d == -1)
            return -1;

        var smallestLargerThanToSwap = 9;
        var e = d+1;
        for (var i = d+1; i < digits.Length; i++)
        {
            //$"Looking at {digits[i]}".Dump();
            if (digits[i] <= digits[d] || digits[i] > digits[e])
                continue;

            smallestLargerThanToSwap = digits[i];
            e = i;
        }
        (digits[d], digits[e]) = (digits[e], digits[d]);
        return digits.Take(d+1).Concat(digits.Skip(d+1).OrderBy(c => c)).ToArray().ToNumber();
    }
}

public static class CodeWarExtensions
{
    public static IEnumerable<int> ToDigits(this long n)
        => n.ToString().ToCharArray().Select(c => int.Parse(c.ToString()));
  
    public static long ToNumber(this int[] digits)
        => long.Parse(string.Join("", digits));
}

File: 4KYU/Next bigger number with the same digits/solution.go
package kata

import "sort"

func Reverse(xs []int) {
	for i, j := 0, len(xs)-1; i < j; i, j = i+1, j-1 {
		xs[i], xs[j] = xs[j], xs[i]
	}
}

func ToDigits(n int) []int {
	var digits []int
	i := 0
	for n != 0 {
		digits = append(digits, n%10)
		n /= 10
		i = i + 1
	}
	Reverse(digits)
	return digits
}

func BuildFinal(xs []int, d int) []int {
	a := xs[0 : d+1]
	b := xs[d+1:]
	sort.Ints(b)
	return append(a, b...)
}

func ToNumber(xs []int) int {
	res := 0
	op := 1
	for i := len(xs) - 1; i >= 0; i-- {
		res += xs[i] * op
		op *= 10
	}
	return res
}

func NextBigger(n int) int {
	digits := ToDigits(n)
	d := -1
	for i := len(digits) - 2; i >= 0; i-- {
		if digits[i] >= digits[i+1] {
			continue
		}
		d = i
		break
	}

	if d == -1 {
		return -1
	}

	e := d + 1

	for i := d + 1; i < len(digits); i++ {
		if digits[i] <= digits[d] || digits[i] > digits[e] {
			continue
		}
		e = i
	}

	t := digits[d]
	digits[d] = digits[e]
	digits[e] = t

	res := ToNumber(BuildFinal(digits, d))

	return res
}
